### DISCUSSION TITLE
Use logger from a router

### DISCUSSION THREAD

**Discussion by metropt** [fastapi/fastapi#7775] [Questions] [answered]
Hi,

I had setup a basic application with fastapi but I can't make logging work. I'm using docker-compose to run the docker: ```command: uvicorn app.main:app --reload --workers 1 --host 0.0.0.0 --port 8000 --log-level debug```

I got the logs from the uvicorn but not from fastapi.

If i run the following line on the main.py where I create the fastapi instance:
```
logger = logging.getLogger('teste')
logger.debug('BLABLA')
app = FastAPI()
```
it outputs to the log:
![image](https://user-images.githubusercontent.com/764040/74266110-a2336180-4cfb-11ea-8a48-e0789afe5fcc.png)

However if I try to log on a router I can't make it output.

**Comment by euri10** [comment, marked as answer]
without a clear example it's impossible to second guess what you're doing, logger works in router at least in this little snippet:
```
import logging

import uvicorn
from fastapi import FastAPI, APIRouter

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)


app = FastAPI()
router = APIRouter()


@app.get("/")
def root():
    logger.debug("root")


@router.get("/")
def routerroot():
    logger.debug("router root")


app.include_router(router, prefix="/router")

if __name__ == '__main__':
    uvicorn.run("l:app", port=8001)
```

```
INFO:     Started server process [17724]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8001 (Press CTRL+C to quit)
DEBUG:    root
INFO:     127.0.0.1:41284 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:41358 - "GET /router HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:41358 - "GET /router/ HTTP/1.1" 200 OK
DEBUG:    router root
```

**Comment by metropt** [comment]
Hi,

I'm using pytest and it was the route of the problem.

Thank you

**Comment by euri10** [comment]
Indeed with pytest you can use -s --log-cli-level=debug and logs won't get
swallowed!

Le mer. 12 févr. 2020 à 12:02 PM, José Xavier <notifications@github.com> a
écrit :

**Comment by tiangolo** [comment]
Thanks @euri10 for the help here! :cake: 

Thanks @metropt  for reporting back and closing the issue :+1:

**Comment by ianyoung** [comment]
I have a similar issue with logging in FastAPI and using the [above code snippet example](https://github.com/tiangolo/fastapi/issues/970#issuecomment-584787196) from @euri10 still doesn't see any logging messages output. I've created a new test application with only a `main.py` file and installed the dependencies with Pipenv:

```
pipenv install fastapi "uvicorn[standard]"
```
Then run the sample snippet:
```
python main.py
```
If I hit both routes I get the following logged to the console:
```
INFO:     Started server process [51541]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:61881 - "GET /router/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:61891 - "GET / HTTP/1.1" 200 OK
```
But no other DEBUG messages at all.

I'm sure it's probably something very simple I'm overlooking but can anyone suggest why I may not be seeing any logging messages?

**Comment by ianyoung** [comment]
Ah. I just hadn't attached a handler to the root logger. This can be achieved with one extra line:
```python
logger.addHandler(logging.StreamHandler())
```
But in the end I created a separate log configuration file where I attached the handlers and formatters in there.
